### PR TITLE
[SPARK-15892][ML] Incorrectly merged AFTAggregator with zero total count

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

Currently, `AFTAggregator` is not being merged correctly. For example, if there is any single empty partition in the data, this creates an `AFTAggregator` with zero total count which causes the exception below:

```
IllegalArgumentException: u'requirement failed: The number of instances should be greater than 0.0, but got 0.'
```

Please see [AFTSurvivalRegression.scala#L573-L575](https://github.com/apache/spark/blob/6ecedf39b44c9acd58cdddf1a31cf11e8e24428c/mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala#L573-L575) as well.

Just to be clear, the python example `aft_survival_regression.py` seems using 5 rows. So, if there exist partitions more than 5, it throws the exception above since it contains empty partitions which results in an incorrectly merged `AFTAggregator`.

Executing `bin/spark-submit examples/src/main/python/ml/aft_survival_regression.py` on a machine with CPUs more than 5 is being failed because it creates tasks with some empty partitions with defualt  configurations (AFAIK, it sets the parallelism level to the number of CPU cores).
## How was this patch tested?

An unit test in `AFTSurvivalRegressionSuite.scala` and manually tested by `bin/spark-submit examples/src/main/python/ml/aft_survival_regression.py`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @jkbradley (the cause was quite deeper than I thought).

**Comment by HyukjinKwon** [comment]
FYI, it passes in branch-1.6 because it does not check whether the count is 0 or not. This checking is introduced in https://github.com/apache/spark/commit/101663f1ae222a919fc40510aa4f2bad22d1be6f.

**Comment by jkbradley** [comment]
@HyukjinKwon Thanks!  The fix looks correct.  Could you please add a tiny unit test which fails before your fix & works afterwards?

**Comment by HyukjinKwon** [comment]
@jkbradley Sure!

**Comment by jkbradley** [comment]
LGTM
Merging with master and branch-2.0
Thank you!

File: mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala [hunk 1]
```diff
@@ -390,6 +390,18 @@ class AFTSurvivalRegressionSuite
     testEstimatorAndModelReadWrite(aft, datasetMultivariate,
       AFTSurvivalRegressionSuite.allParamSettings, checkModelData)
   }
+
+  test("SPARK-15892: Incorrectly merged AFTAggregator with zero total count") {
+    // This `dataset` will contain an empty partition because it has two rows but
+    // the parallelism is bigger than that. Because the issue was about `AFTAggregator`s
+    // being merged incorrectly when it has an empty partition, running the codes below
+    // should not throw an exception.
+    val dataset = spark.createDataFrame(
```
Comment (zzcclp) [inline mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala]: @HyukjinKwon value `spark` is not found here, it should be `sqlContext`, right?

Comment (zzcclp) [inline mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala]: I compile it in branch-1.6.

Comment (zzcclp) [inline mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala]: with branch-2.0, it is OK , i think that this pr should not be merged into branch-1.6 directly.

Comment (HyukjinKwon) [inline mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala]: Oh, it seems this is merged into branch-1.6 too. Yes, it should be `sqlContext` for branch-1.6.

**Comment by HyukjinKwon** [comment]
@zzcclp I made a PR against branch-1.6 here, https://github.com/apache/spark/pull/13630. Thank you for pointing this out quickly.

**Comment by zzcclp** [comment]
@HyukjinKwon OK, but this pr should be reverted first.
@jkbradley  could you revert this pr for branch-1.6 ?

**Comment by HyukjinKwon** [comment]
Yes, it seems it is still failing. I can change the PR to revert this if he is busy for now. Otherwise, I will close mine.

**Comment by zzcclp** [comment]
@HyukjinKwon ,  could you revert this pr for branch-1.6 ?

**Comment by HyukjinKwon** [comment]
@zzcclp I just did here https://github.com/apache/spark/pull/13630

**Comment by zzcclp** [comment]
@HyukjinKwon , just see it, thanks.

**Comment by jkbradley** [comment]
My apologies for the slow response.  I'll revert it now.  Thank you for pinging!

**Comment by jkbradley** [comment]
Reverted commit to branch-1.6 in commit 2f3e327c4cbf163d8536c4451b4829ec7d1886a9

**Pull request closed without merging**